Add some notifications related WKWebsiteDataStoreDelegate calls

beidson opened the pull request:
#### 7e04c00f51164843a92a6ac55aa2917c6442f5f1
<pre>
Add some notifications related WKWebsiteDataStoreDelegate calls
<a href="https://bugs.webkit.org/show_bug.cgi?id=249164">https://bugs.webkit.org/show_bug.cgi?id=249164</a>
rdar://103271182

Reviewed by Ben Nham.

Adds variants of the permissions and showNotification calls to _WKWebSiteDataStoreDelegate

Also fixes a few other issues in the push/notifications code paths.
And adds some logging.

* Source/WebCore/workers/service/ServiceWorkerRegistration.cpp:
(WebCore::ServiceWorkerRegistration::showNotification):
* Source/WebCore/workers/service/server/SWServer.cpp:
(WebCore::SWServer::processPushMessage):
* Source/WebKit/NetworkProcess/Notifications/NetworkNotificationManager.cpp:
(WebKit::NetworkNotificationManager::getPendingPushMessages):
* Source/WebKit/Platform/IPC/DaemonCoders.cpp:
(WebKit::Daemon::Coder&lt;WTF::UUID&gt;::encode): Deleted.
(WebKit::Daemon::Coder&lt;WTF::UUID&gt;::decode): Deleted.
* Source/WebKit/Platform/IPC/DaemonCoders.h:
(WebKit::Daemon::Coder&lt;WTF::UUID&gt;::encode):
(WebKit::Daemon::Coder&lt;WTF::UUID&gt;::decode):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
* Source/WebKit/UIProcess/API/Cocoa/_WKNotificationData.h: Copied from Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreDelegate.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKNotificationData.mm: Copied from Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreDelegate.h.
(-[_WKNotificationData initWithCoreData:dataStore:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKNotificationDataInternal.h: Copied from Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreDelegate.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreDelegate.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::processPushMessage):
* Source/WebKit/UIProcess/Notifications/ServiceWorkerNotificationHandler.cpp:
(WebKit::ServiceWorkerNotificationHandler::showNotification):
* Source/WebKit/UIProcess/Notifications/WebNotificationManagerMessageHandler.cpp:
(WebKit::WebNotificationManagerMessageHandler::showNotification):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.mm:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::initializeNewWebProcess):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::showServiceWorkerNotification):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreClient.h:
(WebKit::WebsiteDataStoreClient::showNotification):
(WebKit::WebsiteDataStoreClient::notificationPermissions):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/PushAPI.mm:
(messageDictionary):
(-[FirePushEventDataStoreDelegate notificationPermissionsForWebsiteDataStore:]):
(-[FirePushEventDataStoreDelegate websiteDataStore:showNotification:]):

Canonical link: <a href="https://commits.webkit.org/257778@main">https://commits.webkit.org/257778@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/340f9cf0b93e98c03d5dd662472f8a3df21d4579

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99980 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9147 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33058 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109327 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169560 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103979 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10032 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/86448 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92425 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107227 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105748 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/7579 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/90861 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34294 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89485 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22252 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2944 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23765 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2904 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/46140 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9037 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43242 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5338 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4752 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

bnham commented on Source/WebKit/UIProcess/API/Cocoa/_WKNotificationData.mm:
```diff
@@ -0,0 +1,53 @@
+    _title = (NSString *)coreData.title;
+    _body = (NSString *)coreData.body;
+    _origin = (NSString *)coreData.originString;
+    _identifier = (NSString *)coreData.notificationID.toString();
```
The ivar assignments need to call `retain` or `copy` on their arguments and there needs to be a `dealloc` implemmentation that `releases` those ivars, since it looks like we compile this file with MRR.

bnham commented on Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
```diff
@@ -129,11 +132,39 @@ void openWindowFromServiceWorker(const String& url, const WebCore::SecurityOrigi
+        return true;
+    }
+
+    HashMap<WTF::String, bool> notificationPermissions() final
```
Seems like we need datastore-aware variants of the various functions that allow the UIProcess to call into WebKit with updated permissions as well (e.g. `WKNotificationManagerProviderDidUpdateNotificationPolicy` and `WKNotificationManagerProviderDidRemoveNotificationPolicies`). Maybe this can be done in a different patch.

bnham reviewed (changes requested):
Mostly seems okay but need to fix the memory management issues.

bnham reviewed (approved):
LGTM

beidson committed:
Add some notifications related WKWebsiteDataStoreDelegate calls
https://bugs.webkit.org/show_bug.cgi?id=249164
rdar://103271182

Reviewed by Ben Nham.

Adds variants of the permissions and showNotification calls to _WKWebSiteDataStoreDelegate

Also fixes a few other issues in the push/notifications code paths.
And adds some logging.

* Source/WebCore/workers/service/ServiceWorkerRegistration.cpp:
(WebCore::ServiceWorkerRegistration::showNotification):
* Source/WebCore/workers/service/server/SWServer.cpp:
(WebCore::SWServer::processPushMessage):
* Source/WebKit/NetworkProcess/Notifications/NetworkNotificationManager.cpp:
(WebKit::NetworkNotificationManager::getPendingPushMessages):
* Source/WebKit/Platform/IPC/DaemonCoders.cpp:
(WebKit::Daemon::Coder<WTF::UUID>::encode): Deleted.
(WebKit::Daemon::Coder<WTF::UUID>::decode): Deleted.
* Source/WebKit/Platform/IPC/DaemonCoders.h:
(WebKit::Daemon::Coder<WTF::UUID>::encode):
(WebKit::Daemon::Coder<WTF::UUID>::decode):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
* Source/WebKit/UIProcess/API/Cocoa/_WKNotificationData.h: Copied from Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreDelegate.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKNotificationData.mm: Copied from Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreDelegate.h.
(-[_WKNotificationData initWithCoreData:dataStore:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKNotificationDataInternal.h: Copied from Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreDelegate.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreDelegate.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::processPushMessage):
* Source/WebKit/UIProcess/Notifications/ServiceWorkerNotificationHandler.cpp:
(WebKit::ServiceWorkerNotificationHandler::showNotification):
* Source/WebKit/UIProcess/Notifications/WebNotificationManagerMessageHandler.cpp:
(WebKit::WebNotificationManagerMessageHandler::showNotification):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.mm:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::initializeNewWebProcess):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::showServiceWorkerNotification):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreClient.h:
(WebKit::WebsiteDataStoreClient::showNotification):
(WebKit::WebsiteDataStoreClient::notificationPermissions):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/PushAPI.mm:
(messageDictionary):
(-[FirePushEventDataStoreDelegate notificationPermissionsForWebsiteDataStore:]):
(-[FirePushEventDataStoreDelegate websiteDataStore:showNotification:]):

Canonical link: https://commits.webkit.org/257778@main
